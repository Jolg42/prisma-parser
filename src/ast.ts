export type Node = DocumentNode | ModelDefinitionNode | FieldDefinitionNode | TypeNode | IdentifierNode | AttributeNode

export type Position = {
  line: number
  column: number
  offset: number
}

export type Location = {
  start: Position
  end: Position
}

export type DocumentNode = {
  kind: 'Document'
  definitions: ModelDefinitionNode[]
  location?: Location
}

export type ModelDefinitionNode = {
  kind: 'ModelDefinition'
  name: IdentifierNode
  fields: FieldDefinitionNode[]
  location?: Location
}

export type FieldDefinitionNode = {
  kind: 'FieldDefinition'
  name: IdentifierNode
  type: TypeNode
  attributes: AttributeNode[]
  location?: Location
}

export type TypeModifier = 'none' | 'array' | 'optional'

export type TypeNode = {
  kind: 'Type'
  name: IdentifierNode
  modifier: TypeModifier
  location?: Location
}

export type IdentifierNode = {
  kind: 'Identifier'
  identifier: string
  location?: Location
}

export type AttributeNode = {
  kind: 'Attribute'
  name: string
  location?: Location
}
