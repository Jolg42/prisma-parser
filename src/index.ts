export { parse } from './parser'
export { print } from './printer'
export { PrismaParsingError } from './prisma-parsing-error'
export * from './ast'
