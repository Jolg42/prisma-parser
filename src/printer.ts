import {
  DocumentNode,
  FieldDefinitionNode,
  Node,
  TypeNode,
  AttributeNode,
  DefinitionNode,
  ModelDefinitionNode,
  ConfigDefinitionNode,
  ConfigOptionNode,
  ExpressionNode,
  BooleanLiteralNode,
  StringLiteralNode,
  FunctionCallNode,
} from './ast'
import { assertNever } from './util'

const TAB_SIZE = 2

type PrintFunction<NodeType extends Node> = (node: NodeType, printerState: PrinterState) => void

function padBy(paddingSize: number, str: string): string {
  return str.padStart(str.length + paddingSize)
}

class PrinterState {
  private indentationLevel = 0
  private lines: string[] = []
  private currentLine = ''

  indent(): this {
    this.indentationLevel++
    return this
  }

  unindent(): this {
    if (this.indentationLevel === 0) {
      throw new Error('Can not unindent beyond 0')
    }
    this.indentationLevel--
    return this
  }

  write(string: string): this {
    this.currentLine += string
    return this
  }

  writeAligned(string: string, offsetFromStartOfTheLine: number): this {
    const paddingSize = Math.max(offsetFromStartOfTheLine - this.currentLine.length, 0)
    this.write(padBy(paddingSize, string))
    return this
  }

  newLine(): this {
    this.lines.push(padBy(this.indentationLevel * TAB_SIZE, this.currentLine))
    this.currentLine = ''
    return this
  }

  writeSeparated<NodeType extends Node>(
    nodes: NodeType[],
    separator: string,
    printNode: PrintFunction<NodeType>,
  ): this {
    if (nodes.length === 0) {
      return this
    }
    printNode(nodes[0], this)
    for (let i = 1; i < nodes.length; i++) {
      if (separator === '\n') {
        this.newLine()
      } else {
        this.write(separator)
      }
      printNode(nodes[i], this)
    }
    return this
  }

  getText(): string {
    if (this.currentLine) {
      this.newLine()
    }
    return this.lines.join('\n')
  }
}

export function print(document: DocumentNode): string {
  const printerState = new PrinterState()
  printDocument(document, printerState)
  return printerState.getText()
}

function printDocument(document: DocumentNode, printerState: PrinterState): void {
  printerState.writeSeparated(document.definitions, '\n', printDefinition)
  printerState.newLine()
}

function printDefinition(definition: DefinitionNode, printerState: PrinterState): void {
  switch (definition.kind) {
    case 'ModelDefinition':
      printModelDefinition(definition, printerState)
      break
    case 'ConfigDefinition':
      printConfigDefinition(definition, printerState)
      break
    default:
      assertNever(definition, 'Unexpected definition type')
  }
}

function printModelDefinition(definition: ModelDefinitionNode, printerState: PrinterState): void {
  const typeOffset = getTypeAlignmentOffset(definition.fields)
  printerState
    .write('model ')
    .write(definition.name.identifier)
    .write(' {')
    .newLine()
    .indent()
    .writeSeparated(definition.fields, '\n', (field) => {
      printField(field, typeOffset, printerState)
    })
    .newLine()
    .unindent()
    .write('}')
    .newLine()
}

function getTypeAlignmentOffset(fields: FieldDefinitionNode[]): number {
  const longestFieldName = Math.max(...fields.map((field) => field.name.identifier.length))
  return longestFieldName + 2 // minimum ident is 2 spaces from field name
}

function printField(field: FieldDefinitionNode, typeOffset: number, printerState: PrinterState): void {
  printerState.write(field.name.identifier).write('  ')
  printType(field.type, typeOffset, printerState)

  for (const attribute of field.attributes) {
    printerState.write(' ')
    printAttribute(attribute, printerState)
  }
}

function printType(type: TypeNode, typeOffset: number, printerState: PrinterState): void {
  printerState.writeAligned(type.name.identifier, typeOffset)
  if (type.modifier === 'array') {
    printerState.write('[]')
  } else if (type.modifier === 'optional') {
    printerState.write('?')
  }
}

function printAttribute(attribute: AttributeNode, printerState: PrinterState): void {
  printerState.write(attribute.name)
}

function printConfigDefinition(definition: ConfigDefinitionNode, printerState: PrinterState): void {
  printerState
    .write(definition.type)
    .write(' ')
    .write(definition.name.identifier)
    .write(' {')
    .newLine()
    .indent()
    .writeSeparated(definition.options, '\n', printConfigOption)
    .newLine()
    .unindent()
    .write('}')
    .newLine()
}

function printConfigOption(option: ConfigOptionNode, printerState: PrinterState): void {
  printerState.write(option.key.identifier).write(' = ')
  printExpression(option.value, printerState)
}

function printExpression(expression: ExpressionNode, printerState: PrinterState): void {
  switch (expression.kind) {
    case 'BooleanLiteral':
      printBooleanLiteral(expression, printerState)
      break
    case 'StringLiteral':
      printStringLiteral(expression, printerState)
      break
    case 'FunctionCall':
      printFunctionCall(expression, printerState)
      break
    default:
      assertNever(expression, 'Unexpected value kind')
  }
}

function printBooleanLiteral(literal: BooleanLiteralNode, printerState: PrinterState): void {
  printerState.write(String(literal.value))
}

function printStringLiteral(literal: StringLiteralNode, printerState: PrinterState): void {
  printerState.write('"').write(literal.value.replace(/"/g, '\\"')).write('"')
}

function printFunctionCall(call: FunctionCallNode, printerState: PrinterState): void {
  printerState.write(call.name.identifier).write('(').writeSeparated(call.arguments, ', ', printExpression).write(')')
}
