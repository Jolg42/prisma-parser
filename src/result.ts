export interface Result<T, Err extends Error> {
  isOk(): boolean
  isError(): boolean

  map<U>(mapper: (value: T) => U): Result<U, Err>
  flatMap<U>(mapper: (value: T) => Result<U, Err>): Result<U, Err>
  orElse(mapper: (error: Err) => Result<T, Err>): Result<T, Err>

  unwrap(): T
}

class ResultOk<T, Err extends Error> implements Result<T, Err> {
  private value: T
  constructor(value: T) {
    this.value = value
  }

  isOk(): boolean {
    return true
  }
  isError(): boolean {
    return false
  }
  map<U>(mapper: (value: T) => U): Result<U, Err> {
    return new ResultOk(mapper(this.value))
  }
  flatMap<U>(mapper: (value: T) => Result<U, Err>): Result<U, Err> {
    return mapper(this.value)
  }

  orElse(): Result<T, Err> {
    return ok(this.value)
  }

  unwrap(): T {
    return this.value
  }
}

class ResultErr<T, Err extends Error> implements Result<T, Err> {
  private error: Err
  constructor(error: Err) {
    this.error = error
  }

  isOk(): boolean {
    return false
  }

  isError(): boolean {
    return true
  }

  map<U>(): Result<U, Err> {
    return new ResultErr(this.error)
  }

  flatMap<U>(): Result<U, Err> {
    return new ResultErr(this.error)
  }

  orElse(mapper: (error: Err) => Result<T, Err>): Result<T, Err> {
    return mapper(this.error)
  }

  unwrap(): T {
    throw this.error
  }
}

export function ok<T, Err extends Error>(value: T): Result<T, Err> {
  return new ResultOk(value)
}

export function err<T, Err extends Error>(error: Err): Result<T, Err> {
  return new ResultErr(error)
}
