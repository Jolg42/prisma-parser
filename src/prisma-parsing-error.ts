import { createCodeFrame } from 'simple-code-frame'
import { Position } from './ast'

export class PrismaParsingError extends SyntaxError {
  position: Position
  source: string

  constructor(message: string, source: string, position: Position) {
    super(message)
    this.name = 'PrismaParsingError'
    this.source = source
    this.position = position
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, PrismaParsingError)
    }
  }

  toString(): string {
    // our line and column numbers are line-based, simple-code-frame ones are 0-based
    const frame = createCodeFrame(this.source, this.position.line - 1, this.position.column - 1, { colors: false })
    return `${this.name}: ${this.message}\n\n${frame}`
  }
}
