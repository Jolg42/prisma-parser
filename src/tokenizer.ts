import { Position } from './ast'
import { PrismaParsingError } from './prisma-parsing-error'

export type TokenType = 'identifier' | 'attribute' | 'string' | 'eof' | '{' | '}' | '(' | ')' | ',' | '[]' | '?' | '='

export type Token = {
  type: TokenType
  token: string
  start: Position
  end: Position
}

type TokenizationRule = {
  type: TokenType
  pattern: RegExp
}

const tokenizationRules: TokenizationRule[] = [
  { type: 'identifier', pattern: /[a-zA-Z][a-zA-Z0-9]*/y },
  { type: 'attribute', pattern: /@[a-zA-Z][a-zA-Z0-9]*/y },
  // eslint is wrong: \" is not an escape sequence, it is supposed to match
  // \ and " characters and \ does not need escaping in this case
  // eslint-disable-next-line no-useless-escape
  { type: 'string', pattern: /"([^"]|\")*"/y },
  { type: '{', pattern: /{/y },
  { type: '}', pattern: /}/y },
  { type: '(', pattern: /\(/y },
  { type: ')', pattern: /\)/y },
  { type: ',', pattern: /,/y },
  { type: '?', pattern: /\?/y },
  { type: '=', pattern: /=/y },
  { type: '[]', pattern: /\[]/y },
]

const SPACE_REGEX = /[\s\n\r]+/y

type LazyTokenizer = () => Token

function createLazyTokenizer(input: string): () => Token {
  let position: Position = {
    offset: 0,
    line: 1,
    column: 1,
  }

  const advancePosition = (byOffset: number): Position => {
    const offset = position.offset + byOffset
    let column = position.column
    let line = position.line
    for (let i = position.offset; i < offset; i++) {
      if (input[i] === '\n') {
        line++
        column = 1
      } else {
        column++
      }
    }

    return {
      offset,
      line,
      column,
    }
  }

  const consumeSpace = () => {
    SPACE_REGEX.lastIndex = position.offset
    const spaceMatch = SPACE_REGEX.exec(input)
    if (spaceMatch) {
      position = advancePosition(spaceMatch[0].length)
    }
  }

  return (): Token => {
    consumeSpace()
    if (position.offset === input.length) {
      return {
        type: 'eof',
        token: '',
        start: position,
        end: position,
      }
    }
    for (const rule of tokenizationRules) {
      rule.pattern.lastIndex = position.offset
      const match = rule.pattern.exec(input)
      if (!match) {
        continue
      }
      const token = match[0]
      const start = position
      const end = advancePosition(token.length)
      position = end

      return {
        type: rule.type,
        token,
        start,
        end,
      }
    }

    throw new PrismaParsingError('Unexpected token', input, position)
  }
}

export class TokenList {
  private currentTokenPointer = 0
  private computedTokens: Token[] = []
  private computeNextToken: LazyTokenizer

  constructor(input: string) {
    this.computeNextToken = createLazyTokenizer(input)
  }

  getPointer(): number {
    return this.currentTokenPointer
  }

  restorePointer(pointer: number): void {
    if (pointer >= this.computedTokens.length) {
      throw new Error('Attempt to set pointer to a not yet computed token')
    }

    this.currentTokenPointer = pointer
  }

  advance(): Token {
    const token = this.peek()
    this.currentTokenPointer++
    return token
  }

  peek(): Token {
    if (this.currentTokenPointer === this.computedTokens.length) {
      this.computedTokens.push(this.computeNextToken())
    }
    return this.computedTokens[this.currentTokenPointer]
  }
}
