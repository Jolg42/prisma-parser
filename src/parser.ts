import { Result, ok, err } from './result'
import { Token, TokenList, TokenType } from './tokenizer'
import { Node, DocumentNode, TypeModifier, Position } from './ast'
import { PrismaParsingError } from './prisma-parsing-error'

export function parse(source: string): DocumentNode {
  return document(new ParserState(source)).unwrap()
}

type ParsingResult<T> = Result<T, PrismaParsingError>
type RuleFunction<ReturnType> = (state: ParserState) => ParsingResult<ReturnType>

class ParserState {
  private tokenList: TokenList
  private source: string

  constructor(source: string) {
    this.source = source
    this.tokenList = new TokenList(source)
  }

  advance(): Token {
    return this.tokenList.advance()
  }

  peek(): Token {
    return this.tokenList.peek()
  }

  getPosition(): number {
    return this.tokenList.getPosition()
  }

  restorePosition(position: number): void {
    return this.tokenList.restorePosition(position)
  }

  error<T>(message: string, position: Position): ParsingResult<T> {
    return err(new PrismaParsingError(message, this.source, position))
  }
}

const attribute = defineNode(token('attribute'), (token) => ({
  kind: 'Attribute',
  name: token.token,
  location: {
    start: token.start,
    end: token.end,
  },
}))

const identifier = defineNode(token('identifier'), (token) => ({
  kind: 'Identifier',
  identifier: token.token,
  location: {
    start: token.start,
    end: token.end,
  },
}))

const typeModifierToken = defineRule<Token | null>((state) => {
  return (optionalToken(state) as ParsingResult<Token | null>).orElse(() => arrayToken(state)).orElse(() => ok(null))
})

const type = defineNode(
  sequence(changeErrorMessage(identifier, 'Expected field type'), typeModifierToken),
  ([name, modifierToken]) => {
    let modifier: TypeModifier = 'none'
    let end = name.location.end
    if (modifierToken) {
      end = modifierToken.end
      modifier = modifierToken?.type === '[]' ? 'array' : 'optional'
    }

    return {
      kind: 'Type',
      name,
      modifier,
      location: {
        start: name.location.start,
        end,
      },
    }
  },
)

const fieldDefinition = defineNode(
  sequence(changeErrorMessage(identifier, 'Expected field name'), type, zeroOrMore(attribute)),
  ([name, type, attributes]) => {
    const end = attributes.length > 0 ? attributes[attributes.length - 1].location.end : type.location.end
    return {
      kind: 'FieldDefinition',
      name,
      type,
      attributes,
      location: {
        start: name.location.start,
        end,
      },
    }
  },
)

const modelDefinition = defineNode(
  sequence(
    keyword('model'),
    changeErrorMessage(identifier, 'Expected model name'),
    token('{'),
    takeUntil(fieldDefinition, '}'),
  ),
  ([modelKeyword, name, , [fields, closeBrace]]) => ({
    kind: 'ModelDefinition',
    name,
    fields,
    location: {
      start: modelKeyword.start,
      end: closeBrace.end,
    },
  }),
)

const document = defineNode(takeUntil(modelDefinition, 'eof'), ([definitions, eof]) => ({
  kind: 'Document',
  definitions,
  location: {
    start: definitions[0]?.location.start ?? 0,
    end: eof.end,
  },
}))

function sequence<Elements extends unknown[]>(
  ...ruleFunctions: { [K in keyof Elements]: RuleFunction<Elements[K]> }
): RuleFunction<Elements> {
  return defineRule((state) => {
    let result = ok([] as unknown[])
    for (const rule of ruleFunctions) {
      result = result.flatMap((elements) => {
        return rule(state).map((nextValue) => [...elements, nextValue])
      })
    }

    return result as ParsingResult<Elements>
  })
}

function takeUntil<ReturnType>(
  elementRule: RuleFunction<ReturnType>,
  tokenType: TokenType,
): RuleFunction<[ReturnType[], Token]> {
  return (state) => {
    let result: ParsingResult<ReturnType[]> = ok([])
    while (state.peek().type !== tokenType && !result.isError()) {
      result = result.flatMap((elements) => {
        return elementRule(state).map((newElement) => [...elements, newElement])
      })
    }
    return result.map((values) => [values, state.advance()])
  }
}

function zeroOrMore<ReturnType>(elementRule: RuleFunction<ReturnType>): RuleFunction<ReturnType[]> {
  return (state) => {
    const result: ReturnType[] = []
    let ruleResult = elementRule(state)
    while (ruleResult.isOk()) {
      result.push(ruleResult.unwrap())
      ruleResult = elementRule(state)
    }
    return ok(result)
  }
}

const identiferToken = token('identifier')
const arrayToken = token('[]')
const optionalToken = token('?')

function keyword(keywordName: string): RuleFunction<Token> {
  return defineRule((state) => {
    return identiferToken(state)
      .orElse((error) => state.error(`Expected keyword ${keywordName}`, error.position))
      .flatMap((token) => {
        if (token.token !== keywordName) {
          return state.error(`Expected keyword ${keywordName}`, token.start)
        }
        return ok(token)
      })
  })
}

function token(tokenType: TokenType): RuleFunction<Token> {
  return defineRule<Token>((state) => {
    const token = state.advance()
    if (token.type === tokenType) {
      return ok(token)
    }
    return state.error(`Expected ${tokenType}`, token.start)
  })
}

function defineNode<T, NodeType extends Node>(
  ruleFn: RuleFunction<T>,
  toNodeFn: (result: T) => NodeType,
): RuleFunction<NodeType> {
  return defineRule((state) => {
    return ruleFn(state).map(toNodeFn)
  })
}

function changeErrorMessage<T>(ruleFn: RuleFunction<T>, message: string): RuleFunction<T> {
  return (state) => {
    return ruleFn(state).orElse((error) => state.error(message, error.position))
  }
}

function defineRule<ReturnType>(ruleFn: RuleFunction<ReturnType>): RuleFunction<ReturnType> {
  return (tokenList) => {
    const position = tokenList.getPosition()
    const result = ruleFn(tokenList)
    if (result.isError()) {
      tokenList.restorePosition(position)
    }
    return result
  }
}
