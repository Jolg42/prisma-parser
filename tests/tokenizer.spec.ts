import { TokenList } from '../src/tokenizer'
import { expectParsingError } from './util'

const documentStart = { offset: 0, line: 1, column: 1 }

const firstLinePosition = (offset: number) => ({ offset, line: 1, column: offset + 1 })
describe('tokenization', () => {
  test('recognizes simple identifier', () => {
    const tokens = new TokenList('foo')
    expect(tokens.advance()).toEqual({
      type: 'identifier',
      token: 'foo',
      start: documentStart,
      end: firstLinePosition(3),
    })
  })

  test('recognizes attributes', () => {
    const tokens = new TokenList('@foo')
    expect(tokens.advance()).toEqual({
      type: 'attribute',
      token: '@foo',
      start: documentStart,
      end: firstLinePosition(4),
    })
  })

  test('recognizes strings', () => {
    const tokens = new TokenList('"some string"')
    expect(tokens.advance()).toEqual({
      type: 'string',
      token: '"some string"',
      start: documentStart,
      end: firstLinePosition(13),
    })
  })

  test('allows to escape quotes inside of a string', () => {
    const tokens = new TokenList('"some \\"string"')
    expect(tokens.advance()).toEqual({
      type: 'string',
      token: '"some \\"string"',
      start: documentStart,
      end: firstLinePosition(15),
    })
  })

  test('recognizes opening curly brace', () => {
    const tokens = new TokenList('{')
    expect(tokens.advance()).toEqual({
      type: '{',
      token: '{',
      start: documentStart,
      end: firstLinePosition(1),
    })
  })

  test('recognizes closing curly brace', () => {
    const tokens = new TokenList('}')
    expect(tokens.advance()).toEqual({
      type: '}',
      token: '}',
      start: documentStart,
      end: firstLinePosition(1),
    })
  })

  test('recognizes square braces', () => {
    const tokens = new TokenList('[]')
    expect(tokens.advance()).toEqual({ type: '[]', token: '[]', start: documentStart, end: firstLinePosition(2) })
  })

  test('recognizes question mark', () => {
    const tokens = new TokenList('?')
    expect(tokens.advance()).toEqual({ type: '?', token: '?', start: documentStart, end: firstLinePosition(1) })
  })

  test('recognizes equals sign', () => {
    const tokens = new TokenList('=')
    expect(tokens.advance()).toEqual({
      type: '=',
      token: '=',
      start: documentStart,
      end: firstLinePosition(1),
    })
  })

  test('ignores spaces', () => {
    const tokens = new TokenList('    foo  ')
    expect(tokens.advance()).toEqual({
      type: 'identifier',
      token: 'foo',
      start: firstLinePosition(4),
      end: firstLinePosition(7),
    })
  })

  test('ignores newlines', () => {
    const tokens = new TokenList('\n\nfoo')
    expect(tokens.advance()).toEqual({
      type: 'identifier',
      token: 'foo',
      start: { offset: 2, line: 3, column: 1 },
      end: { offset: 5, line: 3, column: 4 },
    })
  })

  test('consumes multiple tokens', () => {
    const tokens = new TokenList('foo { bar }')
    expect(tokens.advance().token).toBe('foo')
    expect(tokens.advance().token).toBe('{')
    expect(tokens.advance().token).toBe('bar')
    expect(tokens.advance().token).toBe('}')
  })

  test('returns EOF token when stream is finished', () => {
    const tokens = new TokenList('foo')
    tokens.advance()
    expect(tokens.advance()).toEqual({ type: 'eof', token: '', start: firstLinePosition(3), end: firstLinePosition(3) })
  })

  test('throws error on unexpected token', () => {
    const tokens = new TokenList('ÃŸ')

    expectParsingError(() => tokens.advance())
  })
})

describe('pointer manipulation methods', () => {
  test('.advance advances the pointer', () => {
    const tokens = new TokenList('foo')
    tokens.advance()
    expect(tokens.getPointer()).toBe(1)
  })

  test('.peek returns current token', () => {
    const tokens = new TokenList('foo')
    expect(tokens.peek()).toEqual({ type: 'identifier', token: 'foo', start: documentStart, end: firstLinePosition(3) })
  })

  test('.peek does not advance pointer', () => {
    const tokens = new TokenList('foo')
    tokens.peek()
    expect(tokens.getPointer()).toBe(0)
  })

  test('.restorePointer allows to go back to previous token', () => {
    const tokens = new TokenList('foo')
    tokens.advance()
    tokens.restorePointer(0)
    expect(tokens.advance()).toEqual({
      type: 'identifier',
      token: 'foo',
      start: documentStart,
      end: firstLinePosition(3),
    })
  })

  test('.restorePointer throws if trying to go beyond currently computed tokens', () => {
    const tokens = new TokenList('foo')
    expect(() => tokens.restorePointer(9000)).toThrow('Attempt to set pointer to a not yet computed toke')
  })
})
