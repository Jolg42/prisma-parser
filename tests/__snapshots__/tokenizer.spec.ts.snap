// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tokenization throws error on unexpected token 1`] = `
"PrismaParsingError: Unexpected token

> 1 | ÃŸ
    | ^
"
`;
